diff --git a/ext/mysqlnd/mysqlnd_ps_codec.c b/ext/mysqlnd/mysqlnd_ps_codec.c
index c61bb30edb..a4913f0efb 100644
--- a/ext/mysqlnd/mysqlnd_ps_codec.c
+++ b/ext/mysqlnd/mysqlnd_ps_codec.c
@@ -241,7 +241,11 @@ ps_fetch_time(zval * zv, const MYSQLND_FIELD * const field, const unsigned int p
                t.time_type = MYSQLND_TIMESTAMP_TIME;
        }

-       length = mnd_sprintf(&value, 0, "%s%02u:%02u:%02u", (t.neg ? "-" : ""), t.hour, t.minute, t.second);
+       if (t.second_part > 0) {
+                length = mnd_sprintf(&value, 0, "%s%02u:%02u:%02u.%06u", (t.neg ? "-" : ""), t.hour, t.minute, t.second, t.second_part);
+       } else {
+                length = mnd_sprintf(&value, 0, "%s%02u:%02u:%02u", (t.neg ? "-" : ""), t.hour, t.minute, t.second);
+       }

        DBG_INF_FMT("%s", value);
        ZVAL_STRINGL(zv, value, length);
@@ -322,7 +326,12 @@ ps_fetch_datetime(zval * zv, const MYSQLND_FIELD * const field, const unsigned i
                t.time_type = MYSQLND_TIMESTAMP_DATETIME;
        }

-       length = mnd_sprintf(&value, 0, "%04u-%02u-%02u %02u:%02u:%02u", t.year, t.month, t.day, t.hour, t.minute, t.second);
+
+       if (t.second_part > 0) {
+                length = mnd_sprintf(&value, 0, "%04u-%02u-%02u %02u:%02u:%02u.%06u", t.year, t.month, t.day, t.hour, t.minute, t.second, t.second_part);
+       } else {
+                length = mnd_sprintf(&value, 0, "%04u-%02u-%02u %02u:%02u:%02u", t.year, t.month, t.day, t.hour, t.minute, t.second);
+       }

        DBG_INF_FMT("%s", value);
        ZVAL_STRINGL(zv, value, length);
